analyzers/src/SonarAnalyzer.CSharp/Rules/ReferenceEqualityCheckWhenEqualsExists.cs

```diff
@@ -91,63 +88,35 @@ protected override void Initialize(SonarAnalysisContext context)
                             if (MightOverrideEquals(typeLeft, allInterfacesWithImplementationsOverriddenEquals) ||
                                 MightOverrideEquals(typeRight, allInterfacesWithImplementationsOverriddenEquals))
                             {
-                                c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, binary.OperatorToken.GetLocation()));
+                                c.ReportDiagnosticWhenActive(Diagnostic.Create(Rule, binary.OperatorToken.GetLocation()));
                             }
                         },
                         SyntaxKind.EqualsExpression,
                         SyntaxKind.NotEqualsExpression);
                 });
-        }
 
-        private static bool MightOverrideEquals(ITypeSymbol type, ISet<INamedTypeSymbol> allInterfacesWithImplementationsOverriddenEquals)
-        {
-            return HasEqualsOverride(type) ||
-                allInterfacesWithImplementationsOverriddenEquals.Contains(type) ||
-                HasTypeConstraintsWhichMightOverrideEquals(type, allInterfacesWithImplementationsOverriddenEquals);
-        }
+        private static bool MightOverrideEquals(ITypeSymbol type, ISet<INamedTypeSymbol> allInterfacesWithImplementationsOverriddenEquals) =>
+            HasEqualsOverride(type)
+            || allInterfacesWithImplementationsOverriddenEquals.Contains(type)
+            || HasTypeConstraintsWhichMightOverrideEquals(type, allInterfacesWithImplementationsOverriddenEquals);
 
-        private static bool HasTypeConstraintsWhichMightOverrideEquals(ITypeSymbol type, ISet<INamedTypeSymbol> allInterfacesWithImplementationsOverriddenEquals)
-        {
-            if (type.TypeKind != TypeKind.TypeParameter)
-            {
-                return false;
-            }
+        private static bool HasTypeConstraintsWhichMightOverrideEquals(ITypeSymbol type, ISet<INamedTypeSymbol> allInterfacesWithImplementationsOverriddenEquals) =>
+            type.TypeKind == TypeKind.TypeParameter
+            && type is ITypeParameterSymbol typeParameter
+            && typeParameter.ConstraintTypes.Any(t => MightOverrideEquals(t, allInterfacesWithImplementationsOverriddenEquals));
```

Comment:
_Nitpick:_ Align with current coding style
```suggestion
            && typeParameter.ConstraintTypes.Any(x => MightOverrideEquals(x, allInterfacesWithImplementationsOverriddenEquals));
```